text/0223-cicd-observability-OTEP.md

```diff
@@ -0,0 +1,75 @@
+In addition, oftentimes release issues are not code-based but rather environmental, stemming from issues in the build machines, garbage collection issues of the tool or even a malstructured pipeline step. In order to provide observability into CI/CD environment, especially one with distributed execution mechanism, there's need to monitor various entities such as nodes, queues, jobs and executors (using the Jenkins terms, other tools having respective equivalents, which the specification should abstract with the semantic convention).
+
+The CDF (Continuous Delivery Foundation) has the Events Special Interest Group ([SIG Events](https://github.com/cdfoundation/sig-events)) which explores standardizing on CI/CD event to facilitate interoperability (it is a work-stream within the CDF SIG Interoperability.). The group is working on [CDEvents](https://cdevents.dev/), a standardized event protocol that caters for technology agnostic machine-to-machine communication in CI/CD systems. It makes sense to evaluate alignment between the standards.
+
+OpenTelemetry instrumentation should then support in collecting and emitting the new data. 
+
+OpenTelemetry Collector can then offer designated processors for these payloads, as well as new exporters for designated backend analytics tools, as such prove useful for release engineering needs beyond existing ecosystem.   
+
+## Trade-offs and mitigations
+
+Today’s tools already emit some telemetry, some of which may not easily fit into vendor-agnostic unified semantic conventions. These can be accommodated within extra baggage payload, which may be parsed on a tool-specific fashion. 
+
+## Prior art and alternatives
+
+Today’s tools already emit some telemetry, which can be visualized by the tool’s designated backend, or by general-purpose tools with custom-built dashboards and queries for this specific data. These, however, use proprietary specifications.
+
+## Open questions
+
+Open questions include:
+- Which entity model should be supported to best represent CI/CD domain and pipelines?
+- What are the common CI/CD workflows we aim to support? 
+- What are the primary tools that should be supported with instrumentation in order to gain critical mass on CI/CD coverage?
+- Is CDEvents a good fit of a specification to integrate with? what is the aligmment, overlap and gaps? and if so, how to establish the cross-foundation and cross-group collaboration in an effective manner?
+- How can we bring the existing ecosystem players, both open source and others, to form a concensus and leverage existing knowledge and experience?
+- Which receivers are needed beyond OTLP to support the use cases and workflows?
+- Which exporters are needed to support common backends?
+- Which processors are needed to support the defined workflows?
+
```

Comment:
Sorry for the late answer @horovits
I like to align the two proposals 👍 My PR focuses only on what attributes we should emit on a trace/log that describes a CI/CD pipeline. I think as part of this OTEP we should address the point of @secustor and try to agree on how the trace context can be propagated. This way, different tools can use together and each one contributes spans to the trace so there won't be blind spots.

Reply:
@horovits I agree that shared semantics will help. For a tool to propagate context, it needs to know where to look for it in any incoming trigger or data/policy that is used to decide to perform an action. It also needs how to enrich the context and where to send it to propagate it further. Shared semantics definitely help here, which is very much aligned with the mission of CDEvents.

CDEvents has a simple model for [deployment related of events](https://github.com/cdevents/spec/blob/v0.3.0/continuous-deployment.md) today - it would be great to have shared semantics here as well.  /cc @AloisReitbauer I think this would be of interest to the App Delivery Tag as well.

Reply:
@thschue FYI ^^

Reply:
I would like to retrigger this discussion as I'm facing some decisions which will have an effect on how we are tackling things. 

I see 3 basic ways we should support to propagate context. 

#### HTTP headers
Basically what is already in place the W3C standard. 
The use case would be here that systems get triggered by webhook, e.g. VCS to CI system.

#### Environment variables
I'm imagining here one process handing over context to another. e.g. a pipeline run handing over context to a tool used inside of the run. 

####  Other
Other ways are possible too, tough I would add them to the specification and make them optional e.g. CLI parameters or files containing the necessary information.

Reply:
@secustor  - not sure if you've seen it but there's another OTEP that is specifically related to context propagation at the environment levels that's in progress. https://github.com/open-telemetry/opentelemetry-specification/issues/740

@deejgregor has an working branch for that addition as well that was brought up on today's SIG alongside this OTEP.